### ISSUE TITLE
When chaining ContainsSingle, assertion failed because of FormatException inside FluentAssertion instead of actual difference

### ISSUE THREAD

**Issue by gcelet** [fluentassertions/fluentassertions#1404]
### Description

When chaining `ContainsSingle`, assertion failed because of `FormatException` inside FluentAssertion instead of actual difference.

### Complete minimal example reproducing the issue

```csharp
using FluentAssertions;
using Microsoft.AspNetCore.Mvc.ModelBinding;

	[Test]
	public void Failed_On_FormatException_Inside_FluentAssertions_When_Strings_Are_Different()
	{
		string propertyName = "APropertyName";
		string value = "A value";
		ModelStateDictionary modelState = new ModelStateDictionary();

		modelState.AddModelError(propertyName, $"The value '{value}' is invalid and more characters to be different than expected.");

		modelState.Should().HaveCount(1)
			.And.ContainSingle(kvp => kvp.Key == propertyName)
			.Which.Value.Errors.Should().HaveCount(1)
			.And.ContainSingle(modelError =>
				string.Equals(modelError.ErrorMessage, $"The value '{value}' is invalid."));
	}

	[Test]
	public void Succeed_When_Strings_Are_Same()
	{
		string propertyName = "APropertyName";
		string value = "A value";
		ModelStateDictionary modelState = new ModelStateDictionary();

		modelState.AddModelError(propertyName, $"The value '{value}' is invalid.");

		modelState.Should().HaveCount(1)
			.And.ContainSingle(kvp => kvp.Key == propertyName)
			.Which.Value.Errors.Should().HaveCount(1)
			.And.ContainSingle(modelError =>
				string.Equals(modelError.ErrorMessage, $"The value '{value}' is invalid."));
	}
```

### Expected behavior:

`Failed_On_FormatException_Inside_FluentAssertions_When_Strings_Are_Different` test should failed because actual string is different of expected string.

### Actual behavior:

`Failed_On_FormatException_Inside_FluentAssertions_When_Strings_Are_Different` test failed on `System.FormatException`.

```text
System.FormatException : Index (zero based) must be greater than or equal to zero and less than the size of the argument list.
   at System.Text.StringBuilder.AppendFormatHelper(IFormatProvider provider, String format, ParamsArray args)
   at System.String.FormatHelper(IFormatProvider provider, String format, ParamsArray args)
   at System.String.Format(String format, Object[] args)
   at FluentAssertions.Execution.MessageBuilder.FormatArgumentPlaceholders(String failureMessage, Object[] failureArgs)
   at FluentAssertions.Execution.MessageBuilder.Build(String message, Object[] messageArgs, String reason, ContextDataItems contextData, String identifier, String fallbackIdentifier)
   at FluentAssertions.Execution.AssertionScope.<>c__DisplayClass30_0.<FailWith>b__0()
   at FluentAssertions.Execution.AssertionScope.FailWith(Func`1 failReasonFunc)
   at FluentAssertions.Execution.AssertionScope.FailWith(Func`1 failReasonFunc)
   at FluentAssertions.Execution.AssertionScope.FailWith(String message, Object[] args)
   at FluentAssertions.Collections.SelfReferencingCollectionAssertions`2.ContainSingle(Expression`1 predicate, String because, Object[] becauseArgs)
```

### Versions

* Which version of Fluent Assertions are you using? 
5.10.3

* Which .NET runtime and version are you targeting? 
.NET Core 3.1

### Additional Information

I tried with FluentAssertions 6.0.0-alpha0001: same behavior as 5.10.3

**Comment by jnyrup** [comment]
`ContainSingle` takes an `Expression<Func<T, bool>> predicate`, such that we can extract the `predicate.Body` to provide better failure messages.
At compile time the string interpolation of `{value}` is de-sugared into a `string.Format("The value '{0}' is invalid.", value)`.

So when constructing the failure message, the `{0}` from the `predicate.Body` is wrongly interpreted as an format item.
We include `predicate.Body` a lot of places, so we probably need some escaping of those to guard against this case.

A _workaround_ is to do the string interpolation outside of the predicate.
```cs
string expected = $"The value '{value}' is invalid.";
modelState.Should().HaveCount(1)
    .And.ContainSingle(kvp => kvp.Key == propertyName)
        .Which.Value.Errors.Should().HaveCount(1)
        .And.ContainSingle(modelError => modelError.ErrorMessage == expected);
```

**Comment by Serg046** [comment]
Guess we have the same issue in other places. We can try to find them and then apply the same fix (if the fix is okay, of course).

**Comment by dennisdoomen** [comment]
Yes, that would be very nice indeed.

**Comment by Serg046** [comment]
I tried to look for `string.Format` calls and found tons of them where at least some of them without the issues. Seems it is a task for some existing contributor who knows the current code base.

**Comment by jnyrup** [comment]
@Serg046 Thanks for the hint about `string.Format`!

The `FormatException` is not throw when changing from `ContainSingle(Expression<Func<T, bool>>` to `Contain(Expression<Func<T, bool>>`.

The problem is _double_ formatting - an issue that has haunted us several times before.
The solution is to avoid calling `string.Format` for `expectationPrefix` but only including the constant part
```cs
string expectationPrefix = "Expected {context:collection} to contain a single item matching {0}{reason}, ";
```
and then pass `predicate.Body` on to `FailWith`.

I searched around for other places of pre-formatting `predicate.Body`.
I found a few others places, but I haven't analyzed if they suffer from the same problem.
If they do, I believe they are less likely to be triggered.
Context, the pitfall in `ContainSingle` has existed since its introduction in 1d2e8f56 more than 8 years ago.
```
ConversionSelector.Include
ConversionSelector.Exclude
EventRaisingExtensions.WithArgs
ExcludeMemberByPredicateSelectionRule.ctor
IncludeMemberByPredicateSelectionRule.ctor
PredicateBasedOrderingRule.ctor
```

**Comment by gcelet** [comment]
> A _workaround_ is to do the string interpolation outside of the predicate.
> 
> ```cs
> string expected = $"The value '{value}' is invalid.";
> modelState.Should().HaveCount(1)
>     .And.ContainSingle(kvp => kvp.Key == propertyName)
>         .Which.Value.Errors.Should().HaveCount(1)
>         .And.ContainSingle(modelError => modelError.ErrorMessage == expected);
> ```

@jnyrup Thank you for this workaround: i'm okay using this workaround for the time being.

**Comment by dennisdoomen** [comment]
Should we include the solution to the design guidelines listed in https://github.com/fluentassertions/fluentassertions/issues/1340

**Comment by jnyrup** [comment]
> Should we include the solution to the design guidelines listed in #1340?

About not formatting `predicate.body` or other user-input eagerly?
If yes, then absolutely 👍 
If no, probably also -- I'm just not sure what you're referring to.

**Comment by dennisdoomen** [comment]
To make sure that we capture the preferred solution for rendering an expression in a failure message.

**Comment by tjrobinson** [comment]
I've just come across this myself. I can use the workaround, but is there another issue I can track for if this ever gets resolved? I guess a slightly nicer solution in the short-term would be a more obvious error message when this occurs - with a link to the design principles when they're ready?

**Comment by jnyrup** [comment]
@tjrobinson The problem described here has been resolved in #1406, but it has not yet been released.
It will be released as part of 6.0 which we don't have an ETA of.
To get a notification of when that happens, you can change your Notifications from "Not watching" to "Releases Only".

**Comment by tjrobinson** [comment]
> @tjrobinson The problem described here has been resolved in #1406, but it has not yet been released.

That's great. thanks. Sorry I totally misread "dennisdoomen closed this in #1406 19 days ago" I thought it had just been closed without a fix.
